src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs

```diff
@@ -17,53 +21,68 @@ internal bool TryGetBuffer(in SequencePosition start, in SequencePosition end, o
+            if (type != SequenceType.MultiSegment && startObject != endObject)
                 ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();
 
-            if (type == SequenceType.IMemoryList)
+            if (type == SequenceType.MultiSegment)
             {
-                IMemoryList<T> startMemoryList = (IMemoryList<T>)startSegment;
-                if (startSegment != endSegment)
+                Debug.Assert(startObject is ReadOnlySequenceSegment<T>);
+                ReadOnlySequenceSegment<T> startSegment = Unsafe.As<ReadOnlySequenceSegment<T>>(startObject);
+
+                if (startSegment != endObject)
                 {
-                    next = GetBufferCrossSegment(startIndex, end, ref startMemoryList, ref length);
+                    next = GetBufferCrossSegment(startIndex, end, ref startSegment, ref length);
                 }
 
-                data = startMemoryList.Memory.Slice(startIndex, length);
+                data = startSegment.Memory;
             }
             else if (type == SequenceType.Array)
             {
-                data = new ReadOnlyMemory<T>((T[])startSegment, startIndex, length);
+                Debug.Assert(startObject is T[]);
+
+                data = new ReadOnlyMemory<T>(Unsafe.As<T[]>(startObject));
             }
             else if (type == SequenceType.OwnedMemory)
             {
-                data = ((OwnedMemory<T>)startSegment).Memory.Slice(startIndex, length);
+                Debug.Assert(startObject is OwnedMemory<T>);
+
+                data = (Unsafe.As<OwnedMemory<T>>(startObject)).Memory;
             }
             else if (typeof(T) == typeof(char) && type == SequenceType.String)
             {
-                data = (ReadOnlyMemory<T>)(object)((string)startSegment).AsMemory(startIndex, length);
+                Debug.Assert(startObject is string);
+
+                data = (ReadOnlyMemory<T>)(object)(Unsafe.As<string>(startObject)).AsMemory();
             }
             else
             {
                 data = default;
+                return false;
             }
 
-            return type != SequenceType.Empty;
+            data = data.Slice(startIndex, length);
+            return true;
         }
 
-        private static SequencePosition GetBufferCrossSegment(int startIndex, in SequencePosition end, ref IMemoryList<T> startMemoryList, ref int length)
+        private static SequencePosition GetBufferCrossSegment(int startIndex, in SequencePosition end, ref ReadOnlySequenceSegment<T> startSegment, ref int length)
         {
-            IMemoryList<T> nextSegment = startMemoryList.Next;
-            int currentLength = startMemoryList.Memory.Length - startIndex;
+            Debug.Assert(startSegment != null);
+
+            ReadOnlySequenceSegment<T> nextSegment = startSegment.Next;
+            int currentLength = startSegment.Memory.Length - startIndex;
 
             while (currentLength == 0 && nextSegment != null)
```

Comment:
Add a test for this. Making sure it isn't possible to walk over segments that are past end that are empty.

Reply:
Done